wire-ios-sync-engine/Tests/Source/Calling/MLSConferenceStaleParticipantsRemoverTests.swift

```diff
@@ -0,0 +1,270 @@
+    private let groupID = MLSGroupID.random()
+    private let connectedState = CallParticipantState.connected(
+        videoState: .stopped,
+        microphoneState: .muted
+    )
+
+    private var mlsService: MockMLSService!
+    private var sut: MLSConferenceStaleParticipantsRemover!
+
+    override func setUp() {
+        super.setUp()
+        mlsService = MockMLSService()
+        sut = MLSConferenceStaleParticipantsRemover(
+            mlsService: mlsService,
+            context: uiMOC,
+            removalTimeout: 0.4
+        )
+
+        let selfUser = ZMUser.selfUser(in: uiMOC)
+        selfUser.remoteIdentifier = UUID()
+    }
+
+    override func tearDown() {
+        mlsService = nil
+        sut = nil
+        super.tearDown()
+    }
+
+    func test_ItRemovesStaleParticipantsAfterTimeout() {
+        // GIVEN
+
+        // create call participants
+        let participants = [
+            createMLSParticipant(state: .connecting),
+            createMLSParticipant(state: .connecting),
+            createMLSParticipant(state: connectedState),
+            createMLSParticipant(state: connectedState)
+        ]
+
+        // mock subconversation members
+        mlsService.mockSubconversationMembers = { _ in
+            return participants.map(\.mlsClientID)
+        }
+
+        // set expectations
+        let expectations = expectations(from: participants)
+        mlsService.mockRemoveMembersFromConversation = { clientIDs, _ in
+            guard let id = clientIDs.first else { return }
+            expectations[id]?.fulfill()
+        }
+
+        // WHEN
+        _ = sut.receive(
+            MLSConferenceParticipantsInfo(
+                participants: participants.map(\.callParticipant),
+                subconversationID: groupID
+            )
+        )
+
+        // THEN
+        wait(for: Array(expectations.values), timeout: 0.5)
+    }
+
+    func test_ItDoesntRemoveParticipantsThatReconnectedBeforeTimeout() {
+        // GIVEN
+
+        // create call participants
+        let participants = [
+            createMLSParticipant(state: .connecting),
+            createMLSParticipant(state: .connecting),
+            createMLSParticipant(state: .connecting)
+        ]
+
+        // mock subconversation members
+        mlsService.mockSubconversationMembers = { _ in
+            return participants.map(\.mlsClientID)
+        }
+
+        // WHEN
+        _ = sut.receive(
+            MLSConferenceParticipantsInfo(
+                participants: participants.map(\.callParticipant),
+                subconversationID: groupID
+            )
+        )
+
+        participants[0].updateState(connectedState)
+        participants[1].updateState(connectedState)
+
+        let expectations = expectations(from: participants)
+        mlsService.mockRemoveMembersFromConversation = { clientIDs, _ in
+            guard let id = clientIDs.first else { return }
+            expectations[id]?.fulfill()
```

Comment:
Why do we expect the members to be removed when the test name suggests otherwise?

Reply:
I see now it's configured in the helper method

Reply:
We expect the participants that reconnected not to be removed. The helper method `expectations(from: participants)` creates expectations, for the participants that should not be removed, the expectations will be inverted. If `removeMembersFromConversation` is called for participants that shouldn't be removed, the expectations will be fulfilled and it will result in the test failing.